### PR TITLE
Refactor codes to improve lighthouse score

### LINKED ISSUE #20: feat: fix errors reported by Google Chrome's Lighthouse to improve its score
We should get all 100's here:

image

### PR DESCRIPTION
**Pull request by jateez**
Before:
image

After:
image

## Description
This pull request fixes errors reported by Google Chrome's Lighthouse to improve performance, accessibility, and best practices scores. Key changes include optimizing error handling, improving image handling, and refactoring JavaScript functionality.

resolves #20

### Key Changes:
In app.js:
- Added error handling in API-related functions.
- Used Promise.all for faster loading of contributors.
- Replaced jQuery scroll-to-top with a native JS version.

In index.html:
- Optimized images using <picture> with WebP.
- Improved accessibility with aria-label attributes.
- Changed text from "See More" to "See repository" to increase SEO scores

Additional:
- ~~Preloaded CSS for better performance.~~

## Limitations

The performance and best practices scores are impacted by external factors such as deployments and GitHub API rate limits.
The remaining accessibility issues are related to color contrast in the text on the navigation bar, which affects design accessibility.

## Checklist

- [x] my PR is focused and contains one wholistic change
- [x] I have added screenshots or screen recordings to show the changes

### CODE REVIEW

**Comment by jateez** [comment]
I made several changes to improve the Lighthouse scores. If there are any aspects of these updates that you feel need adjustment or don't align with the project’s goals, please feel free to provide feedback, and I’ll be happy to make any necessary fixes or adjustments.

File: img/3d_line.webp
Comment (yurijmikhalevich) [inline img/3d_line.webp:1]: Thank you for converting the images 🎉  Can you please remove the original, `png` ones?

File: index.html [hunk 1]
```diff
@@ -24,7 +9,8 @@
     <meta property="og:description" content="A community of software engineers and product developers, originally created for students of KubSAU">
     <meta property="og:image" content="https://github.com/move-fast-and-break-things/mfabt_site/assets/92587254/18e628c2-26bb-4f31-8a7f-5173e218ed40">
     <meta property="og:url" content="https://movefastandbreakthings.club/">
-    <link rel="stylesheet" href="main.css">
+    <link rel="preload" href="main.css" as="style" onload="this.onload=null;this.rel='stylesheet'">
+    <noscript><link rel="stylesheet" href="main.css"></noscript>
```
Comment (yurijmikhalevich) [inline index.html:13]: Does this have any impact on our case?

If not, let's revert it to the original style declaration for simplicity 😄

File: index.html [hunk 2]
```diff
@@ -68,7 +54,10 @@ <h1>
           </div>
 
           <div class="line_3d">
-            <img src="img/3d_line.png" alt="">
+            <picture>
+              <source srcset="img/3d_line.webp" type="image/webp">
+              <img src="img/3d_line.png" alt="">
+            </picture>
```
Comment (yurijmikhalevich) [inline index.html:60]: It's ok to remove the "png" images completely. [WebP is supported by all major browsers](https://caniuse.com/?search=webp) and is ok to use as the only option.

image

File: index.html [hunk 3]
```diff
@@ -1,20 +1,5 @@
 <!DOCTYPE html>
 <html lang="en">
-  <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.7.1/jquery.min.js"></script>
```
Comment (yurijmikhalevich) [inline index.html:3]: I like that we get rid of jquery. Thank you 🔥

File: app.js [hunk 4]
```diff
@@ -1,97 +1,140 @@
...
+      return {
+        name: userData.name || userData.login,
+        bio: userData.bio,
+        avatar: userData.avatar_url,
+        html_url: userData.html_url,
+      };
+    }));
+  } catch (error) {
+    console.error("Error fetching authors:", error);
+    return [];
```
Comment (yurijmikhalevich) [inline app.js:55]: let's let this one crash. if fetching fails, it fails. let the outer code handle it
```suggestion
  const response = await fetch(apiURL);
  if (!response.ok) {
    throw new Error(`HTTP error! status: ${response.status}`);
  }
  const data = await response.json();
  return Promise.all(data.map(async (element) => {
    const userResponse = await fetch(element.url);
    if (!userResponse.ok) {
      throw new Error(`HTTP error! status: ${userResponse.status}`);
    }
    const userData = await userResponse.json();
    return {
      name: userData.name || userData.login,
      bio: userData.bio,
      avatar: userData.avatar_url,
      html_url: userData.html_url,
    };
  }));
```

File: app.js [hunk 5]
```diff
@@ -1,97 +1,140 @@
...
+                <div class="name_and_img">
+                    <img src="${author.avatar}" alt="${author.name}'s avatar">
+                    <p class="person_name">${author.name}</p>
+                </div>
+                <p class="about_person">${author.bio || 'No biography available'}</p>
+            `;
+      container.appendChild(personDiv);
+    });
+  } catch (error) {
+    console.error("Error setting authors:", error);
```
Comment (yurijmikhalevich) [inline app.js:80]: for this function, because we aren't doing anything besides the error handling here, let's remove the `try...catch` to and add error handling when we call it?

File: app.js [hunk 6]
```diff
@@ -1,97 +1,140 @@
...
+      top: 0,
+      behavior: 'smooth'
+    });
+  });
 }
 
-window.onload = function () {
+// Wait for DOM content to be loaded before executing scripts
+document.addEventListener('DOMContentLoaded', () => {
   setAuthors(members_url);
```
Comment (yurijmikhalevich) [inline app.js:137]: this is a follow-up to [that](https://github.com/move-fast-and-break-things/mfbt.community/pull/59/files?diff=split&w=1#r1797626690) comment above

like this
```suggestion
  setAuthors(members_url).catch(console.error);
```

File: app.js [hunk 7]
```diff
@@ -1,97 +1,140 @@
...
+                <span class="developer-name">
+                    <a href="${contributor.html_url}" class="tooltip" target="_blank">  
+                        <img src="${contributor.avatar}" alt="${contributor.name}'s avatar" class="shining-image">  
+                        <span class="tooltip-text">${contributor.name}</span>  
+                    </a>
+                </span>
+            `).join('');
+  } catch (error) {
+    console.error("Error setting contributors:", error);
+    element.innerHTML = "Error loading developers";
```
Comment (yurijmikhalevich) [inline app.js:99]: this is good. Thank you 🙌 

let's make the message more user-friendly

```suggestion
    element.innerHTML = "Can't load the developer list. Check out the repository page to see all of the contributors 😅";
```

**Review by yurijmikhalevich — COMMENTED** [review]
Thank you very much for doing this 🔥 

This is a great PR with a clear impact.

Can you please address a few things I have highlighted in the comments and let's get this merged 😄

File: index.html (hunk 1 above)
Comment (jateez) [reply index.html]: initially i thought it was needed to increase lighthouse score on performance aspect, but when i tested it again it doesn't really increase the score. I reverted it to the original style

**Comment by jateez** [comment]
Hi, sorry for the late responses. I was a bit busy with my exams. 

Thank you for the feedbacks. I adjusted the code according to your review. kindly check it again if you have the time.

**Review by yurijmikhalevich — APPROVED** [review]
This is great! Thank you 🔥 

Ideally, we would add `alt` text to the images, but we can do this in a follow up.
